### PR TITLE
C++: Correctly handle destructors at the end of range-based for-loops in the IR

### PR DESCRIPTION
**Pull request by jketema**

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 1]
```diff
@@ -1216,6 +1218,19 @@ class TranslatedRangeBasedForStmt extends TranslatedStmt, ConditionContext {
...
+      child = this.getChild(destructorId) and
+      result = this.getChild(destructorId + 1).getFirstInstruction(kind)
+    )
+    or
+    exists(int lastDestructorIndex |
+      lastDestructorIndex =
+        max(int n | exists(this.getChild(n)) and n >= this.getFirstDestructorCallIndex()) and
+      child = this.getChild(lastDestructorIndex) and
+      result = this.getParent().getChildSuccessor(this, kind)
+    )
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll:1233]: I basically did the same thing as for for-loops. What I don't quite understand is why this case is needed. Before this change `result = this.getParent().getChildSuccessor(this, kind)` did not occur in this predicate and still everything worked.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll:1233]: This was handled by `result = this.getParent().getChildSuccessor(this, kind)` in `getChildFalseSuccessor` before your change. That is, the last element would be the `this.getCondition()`, and from there we'd have a false successor to `this.getParent().getChildSuccessor(this, kind)`.

Now that you're adding the destructor call after the false edge we need to make sure that we actually step to the parent after those destructor calls (as you're correctly doing here).

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll:1233]: Hm, yeah now that you mention it. We only call `getLastChild` when it also holds that we don't handle destructors explicitly.

`getLastChild` is invoked recursively through the other overrides of that predicate, though. So I think I still need to convince myself that we can remove it safely in this case, though 🤔

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll:1233]: In any case, the current implementation matches the one in `ForStmt` (via inheritance of `LoopStmt`), which also has explicit destructor handling.

**Review by MathiasVP — APPROVED** [review]
LGTM if DCA comes back happy!

**Comment by jketema** [comment]
DCA looks ok, except that putty and bochs failed. Don't think it's worth re-running for those.

**Pull request merged**